### ISSUE TITLE
Passing custom Configurations to TestNGRunner

### ISSUE THREAD

**Issue by GayanSandaruwan** [cucumber/cucumber-jvm#2349]
Currently the TestNGCucumberRunner is final. So we can't add custom behavior.

We have different glue paths and feature paths for each test suites. In io.cucumber we achieved this through by adding two parameters in the testNG.xml and then customize the TestNGCucumberRunner to with new Cucumber Options on each suite run dynamically. 
Since the tests are being run in parallel we cant pass them via System Properties or by setting annotations dynamically to be used in the Runner.

#2078 is also referring something similar for Cucumber Junit  runner.

**Comment by mpkorstanje** [comment]
Extension is something that should be designed for. This ensures the extension points can be documented and tested. This in turn ensures we don't break an extension point by accident.

Could you elaborate on how you are passing options from testng.xml to the runner?

**Comment by GayanSandaruwan** [comment]
Our goal is to run the test suites in parallel. ( not methods, not scenarios, not features). And these test suites are different packages in the feature path. Since we have thousands of test cases, it's hard to pass filter tags differently. We use a common tag which specify each module. ( we have around 20 likewise tags )
So in info.cukes version we achieved this via adding two parameters, gluePaths, featurePaths in testNG.xml, then duplicated the Runner to create our own. There we initialized the runner with new cucumber Options for glue and feature taken from the test Suite xml parameters.
From the way current design is, the only option was to create different classes with different annotations. But I think given the context you might have realized that its hard to do.

So we solved this this time as well by duplicating the Runner, and passing cutomOptions in constructor. This was not neat a neat solution.
Let me know if I'm not clear.

**Comment by GayanSandaruwan** [comment]
So, if it's not possible to allow extending, I think it would be nice to allow custom cucumber options to be passed to the Runner. I can send a pull request if you are interested.

**Comment by mpkorstanje** [comment]
I think it might be useful in general for Cucumber to take parameters from TestNG.xml rather then only the annotation.

But I would like to understand how this works a bit better before asking for a pull request.

> So in info.cukes version we achieved this via adding two parameters, gluePaths, featurePaths in testNG.xml, then duplicated the Runner to create our own. There we initialized the runner with new cucumber Options for glue and feature taken from the test Suite xml parameters.

I do not use TestNG on a regular basis. Could you show me an example of what you've described here?

**Comment by GayanSandaruwan** [comment]
Following is a sample testNG.xml suite that I'm using.
`<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Test Suite">
	<parameter name="cucumber.filter.tags.file" value="CustomTags.txt" />

 	<test name="Load Operation Apps" junit="false" preserve-order="true">
		<parameter name="cucumber.features" value="test/feature/estateManager/loadOperationApps"/>
		<parameter name="cucumber.glue" value="com.abccompany.web.stepDefinitions"/>
		<classes>
			<class name="com.abccompany.core.automation.cucumber.RunCucumberTests"/>
		</classes>
	</test>
 	
	<test name="Customer Credit Account Withdrawal Reason Maintenance" junit="false" preserve-order="true">
		<parameter name="cucumber.features" value="test/feature/estateManager/reasonMaintenance/customerCreditAccountWithdrawalReason"/>
		<parameter name="cucumber.glue" value="com.abccompany.web.stepDefinitions"/>
		<classes>
			<class name="com.abccompany.core.automation.cucumber.RunCucumberTests"/>
		</classes>
	</test>
	
 	<test name="Alert Maintenance" junit="false" preserve-order="true">
		<parameter name="cucumber.features" value="test/feature/estateManager/alertMaintenance"/>
		<parameter name="cucumber.glue" value="com.abccompany.web.stepDefinitions"/>
		<classes>
			<class name="com.abccompany.core.automation.cucumber.RunCucumberTests"/>
		</classes>
	</test>
	
 	<test name="Alert Type Maintenance" junit="false" preserve-order="true">
		<parameter name="cucumber.features" value="test/feature/estateManager/alertTypeMaintenance"/>
		<parameter name="cucumber.glue" value="com.abccompany.web.stepDefinitions"/>
		<classes>
			<class name="com.abccompany.core.automation.cucumber.RunCucumberTests"/>
		</classes>
	</test>
	
	<test name="Attribute Option Set Maintenance" junit="false" preserve-order="true">
		<parameter name="cucumber.features" value="test/feature/estateManager/attributeOptionSetMaintenance"/>
		<parameter name="cucumber.glue" value="com.abccompany.web.stepDefinitions"/>
		<classes>
			<class name="com.abccompany.core.automation.cucumber.RunCucumberTests"/>
		</classes>
	</test>
</suite>`

The cucumber.tags.file is a parameter I introduced in our custom runner which would take the tags from a file.

In runtime, when the Runner is initialized I'm passing the customOptions made with the extra options taken from suite xml, to the runner.

In a side note :   I believe this testNG xml reading part should somehow handled in Runner itself, not passed from out but I wasnt able to access the  TestNGContext from  the runner ( could be possible, but could figure out how)

**Comment by GayanSandaruwan** [comment]
Given you seem more interested in enabling the Runner to get from suite parameters rather than passing custom Options to the runner as a second parameter.
And if you are interested in making this happen
Then if you can point me a way to access testNGContext within the Runner
Then I'll be able to submit a PR

**Comment by mpkorstanje** [comment]
I would first like to understand how parameters work.

Passing custom options, or an options provider is another solution but we'd have to think about the way it interacts with the other ways cucumber accepts options.

**Comment by GayanSandaruwan** [comment]
In abstract, when the testNG is executing each suite, the parameters for that suite is accessible through context. 

Currently in the TestNGCucumberRunner the options are passed in the following order.

1. propertiesFileOptions
2. annotationOptions
3. environmentOptions
4. runtimeOptions

options are overridden in the order. My suggestion is to add the testNG suite parameters/ custom options at 3rd. so will be following order.

1. propertiesFileOptions
2. annotationOptions
3. customOptions
3. environmentOptions
4. runtimeOptions

**Comment by mpkorstanje** [comment]
That makes sense. So you'd use something like this:

```java
public interface ConfigurationParameters {

	Optional<String> get(String key);

}
```

```java
    @BeforeClass(alwaysRun = true)
    @Parameters({ "cucumber.glue", "cucumber.features" })
    public void setUpClass(String glue String features) {
        ConfigurationParameters parameters = key -> {
            if ("cucumber.glue".equals(key)) {
                return Optional.ofNullable(glue);
            }
            if ("cucumber.features".equals(key)) {
                return Optional.ofNullable(features);
            }
            return Optional.empty();
        };
        testNGCucumberRunner = new TestNGCucumberRunner(this.getClass(), parameters);
    }
```    

This looks doable. The `CucumberPropertiesParser` could be changed to accept the `ConfigurationParameters` as an argument in addition to a `Map<String, String>`.

Would you like to draft a PR for this? To be accepted the PR should include the code, tests and some javadoc and a section in the `testng/README.md` that explains how to use this feature.

**Comment by mpkorstanje** [comment]
I also wonder if we can make this _just work_ by reading all TestNG parameters at once in the  `AbstractTestNGCucumberTests`.

**Comment by GayanSandaruwan** [comment]
IMO, in a parallel environment, reading all the parameters at once will not be a good way to do it.

**Comment by mpkorstanje** [comment]
Could you explain how you would read all parameters and how it would fail? 

From looking at the documentation it does look like `ITestContext` should be safe to use concurrently:

```java
    @BeforeClass(alwaysRun = true)
    public void setUpClass(ITestContext context) {
        XmlTest currentXmlTest = context.getCurrentXmlTest();
        ConfigurationParameters parameters = currentXmlTest::getParameter;
        testNGCucumberRunner = new TestNGCucumberRunner(this.getClass(), parameters);
    }
``

```java
public interface ConfigurationParameters {
	String get(String key);
}  
```

**Comment by GayanSandaruwan** [comment]
Okay I misunderstood what you asked. Yes it is better to read all the parameters for the suite.

sample would look likr

in RunCucumberTests
```
	@BeforeClass(alwaysRun = true)
	public void setUpClass(ITestContext context) {

		List<String, String> parameters = context.getCurrentXmlTest().getAllParameters();

		testNGCucumberRunner = new CucumberTestNGRunner(this.getClass(), parameters);

	}
```
In runner 

```
class TestNGCucumberRunner(Class<?> clazz, Map<String, String> additionalOptions) {

		RuntimeOptions propertiesFileOptions = new CucumberPropertiesParser()
				.parse(CucumberProperties.fromPropertiesFile())
				.build();

		RuntimeOptions annotationOptions = new CucumberOptionsAnnotationParser()
				.withOptionsProvider(new TestNGCucumberOptionsProvider())
				.parse(clazz)
				.build(propertiesFileOptions);


		RuntimeOptions otherOptions = new CucumberPropertiesParser()
				.parse(additionalOptions)
				.build(annotationOptions);

		RuntimeOptions environmentOptions = new CucumberPropertiesParser()
				.parse(CucumberProperties.fromEnvironment())
				.build(otherOptions);

		RuntimeOptions runtimeOptions = new CucumberPropertiesParser()
				.parse(CucumberProperties.fromSystemProperties())
				.enablePublishPlugin().build(environmentOptions);
                        ..............................
```

**Comment by GayanSandaruwan** [comment]
I agree with you but couldn't find a suitable method in CucumberPropertiesParser to parse properties from a CustomType.
I'll have to 
Get all the parameters from the TestNG as Map and then converting them to ConfugurationProperties type
To use it with CucumberPropertiesParser then convert it back to a Map

This feels like a bit of a over kill. Please advice.

**Comment by mpkorstanje** [comment]
Lets do it with a map first. We can refactor that afterwards.

**Comment by GayanSandaruwan** [comment]
@mpkorstanje  I have sent the Pull request, kindly take a look at it and let me know.

**Comment by GayanSandaruwan** [comment]
I was trying to send a updated pull request back and noticed that, you have already implemented and merged ?

**Comment by mpkorstanje** [comment]
I've pushed an additional commit to your `main` branch. The commit doesn't exists in the `main` branch of this repository.

**Comment by mpkorstanje** [comment]
Fixed by db3e17b1e5060c4f4fae6b9553cd19336a66ce99 and 03ea225d32cfc9fb986f1a08834ae5340947018d.

**Comment by aniljain1982** [comment]
TestNG:
    ---------
    <?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
    <suite name="Sauce Demo Cucumber Test Automation">
        <parameter name="browser" value="Chrome" />
        <test thread-count="2" name="test" parallel="classes">
            <classes>
                <class name="com.automation.cucumber.runner.ui.SauceDemoRunner" />
            </classes>
        </test> 
    </suite> 
    
    Step definitions base class:
----------------------------
    import org.openqa.selenium.WebDriver;
    import org.testng.Reporter;
    import org.testng.xml.XmlTest;
    
    public class BaseStepDefinition {
        WebDriver driver;
        XmlTest xmlTest;
    
        public BaseStepDefinition() {
            try {
                xmlTest = Reporter.getCurrentTestResult().getTestContext().getCurrentXmlTest();
                System.out.println(xmlTest.getParameter("browser"));
            } catch (Exception e) {
                e.printStackTrace();
            }
    
        }
    
    }
    
    Step definitions class:
--------------------------------
    import io.cucumber.java.After;
    import io.cucumber.java.en.Given;
    
    public class SauceDemoStepDefinition extends BaseStepDefinition {
    
        @After
        public void tearDown() throws Exception {
    
        }
    
        @Given("^User is on login page$")
        public void user_is_on_login_page() throws Exception {
    
        }
    
    }
